src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -408,24 +413,26 @@ public unsafe bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value
                             }
 
                             i = entry.next;
-                            if (collisionCount >= entries.Length)
+
+                            allowedCollisions--;
+                            if (allowedCollisions < 0)
                             {
                                 // The chain of entries forms a loop; which means a concurrent update has happened.
                                 // Break out of the loop and throw, rather than looping forever.
                                 goto ConcurrentOperation;
                             }
-                            collisionCount++;
-                        }
+                        } while (true);
```

Comment:
Would this generate worse code, because it certainly is clearer (replacing lines 416-424)?

```
} while (allowedCollisions-- > 0);

// The chain of entries forms a loop; which means a concurrent update has happened.
// Break out of the loop and throw, rather than looping forever
goto ConcurrentOperation;
```

Reply:
Yes, that's nice :) `while (true)` never sits well with me.